scipy/optimize/_differentialevolution.py

```diff
@@ -1581,10 +1581,8 @@ def _select_samples(self, candidate, number_samples):
         obtain random integers from range(self.num_population_members),
         without replacement. You can't have the original candidate either.
         """
-        idxs = list(range(self.num_population_members))
-        idxs.remove(candidate)
-        self.random_number_generator.shuffle(idxs)
-        idxs = idxs[:number_samples]
+        idxs = self.random_number_generator.choice(self.num_population_members - 1, number_samples, replace=False)
+        idxs = idxs + (idxs >= candidate)
```

Comment:
It's not clear to me what this line does. It's important in the sampling process that the candidate solution is not a member of `idxs`. Can you explain the line in detail?

Reply:
In the initial choice, `candidate`, `candidate + 1`, ..., `num_population_members - 2` are used to represent integers `candidate + 1`, `candidate + 2`, ..., `num_population_members - 1`. This shifts them all so that the final values don't include `candidate` and so that `num_population_members - 1` is included. This was not as obvious as the original solution, which is is why I asked for a comment in the code explaining how it works.

Reply:
yes, that's correct, I'll add in some comments